librbd: new API method to force break a peer's exclusive lock

dillaman opened the pull request.

trociny commented:
@dillaman I assume you want to merge this before exclusive lock refactoring (managed lock) from rbd-mirror-ha branch?

trociny commented:
I have a similar implementation for "get lock owner" and "break lock" requests, for rbd-mirror-ha branch [1]. A difference is that I thought it would be more useful to check if the watcher is alive in "get locker" request, not in "break locker" request (returning -ENOTCONN if it is dead), so a user could use "get locker" request to get full locker status. Actually I think there is no much difference though.

[1] https://github.com/trociny/ceph/commit/0a6fbc8563ffb2dbd0ecf8fdd05b69f73d7e98e0

trociny commented on src/librbd/exclusive_lock/AcquireRequest.cc:
```diff
@@ -397,79 +398,30 @@ void AcquireRequest<I>::send_get_lockers() {
+    AcquireRequest<I>, &AcquireRequest<I>::handle_get_lockers>(this);
+  auto req = GetLockerRequest<I>::create(m_image_ctx, &m_locker, ctx);
+  req->send();
 }
```
nit: After this change I would rename GET_LOCKERS state to GET_LOCKER (and the method to send_get_locker).

dillaman commented:
@trociny Yeah, I think the changes in the wip-rbd-mirror-ha are too high risk for a backport all the way back to jewel.

trociny commented:
@dillaman I am thinking how useful it is in general case. I expect it works only if you have two clients, one of them misbehaves and you want to force another one. If you have 3 clients, and the lock owner misbehaves, there are chances that 2 other clients run forced lock acquire simultaneously, and one of them blacklists another one that just acquired the lock before. For rbd-mirror-HA leader locker, I (am trying to) solve this by providing "get owner" and "break lock" methods to rbd-mirror, so it calls "break lock" only for an owner it already knows and believes that it misbehaves.

dillaman commented:
@trociny I am going to tweak this to expose a "get_lock_owner" / "break_lock_owner" API. I really wanted to avoid exposing the details of locking to make integration simpler, but I think you are correct that in order to cover all possible corner cases the client will potentially need better control.

dillaman commented:
@trociny New approach is now posted

trociny commented on src/librbd/exclusive_lock/BreakRequest.cc:
```diff
@@ -0,0 +1,184 @@
+
+template <typename I>
+void BreakRequest<I>::send_blacklist() {
+  if (!m_image_ctx.blacklist_on_break_lock) {
```
minor: In future we will need blacklist_on_break_lock as a parameter for BreakRequest::create. So it could be made from the start (otherwise when this is merged to generic lock implementation).

trociny commented on src/test/pybind/test_rbd.py:
```diff
@@ -1243,6 +1245,36 @@ def test_follower_write(self):
+
+                    image.lock_acquire(RBD_LOCK_MODE_EXCLUSIVE)
+                    assert_raises(ConnectionShutdown,
+                                  blacklist_image.lock_release())
```
@dillaman The parenthesis after blacklist_image.lock_release look wrong. The function name should be passed, not the result of its call.

And fixing this reveals that actually ConnectionShutdown is not raised for me because ReleaseRequest::handle_unlock clears the return value:
```
2017-01-04 17:53:54.013369 7f28467fc700 10 librbd::exclusive_lock::ReleaseRequest: send_close_object_map
2017-01-04 17:53:54.013371 7f28467fc700 10 librbd::object_map::UnlockRequest: 0x7f2854000cf0 send_unlock: oid=rbd_object_map.17e02ae8944a
2017-01-04 17:53:54.013814 7f28597fa700 10 librbd::object_map::UnlockRequest: 0x7f2854000cf0 handle_unlock: r=-108
2017-01-04 17:53:54.013820 7f28597fa700 -1 librbd::object_map::UnlockRequest: failed to release object map lock: (108) Cannot send after transport endpoint shutdown
2017-01-04 17:53:54.013825 7f28597fa700 10 librbd::exclusive_lock::ReleaseRequest: handle_close_object_map: r=0
2017-01-04 17:53:54.013828 7f28597fa700 10 librbd::exclusive_lock::ReleaseRequest: send_unlock
2017-01-04 17:53:54.013829 7f28597fa700 10 librbd::ExclusiveLock: 0x7f28500187b0 handle_releasing_lock
2017-01-04 17:53:54.014153 7f28597fa700 10 librbd::exclusive_lock::ReleaseRequest: handle_unlock: r=-108
2017-01-04 17:53:54.014157 7f28597fa700 -1 librbd::exclusive_lock::ReleaseRequest: failed to unlock: (108) Cannot send after transport endpoint shutdown
2017-01-04 17:53:54.014163 7f28597fa700 10 librbd::ImageState: 0x2104a00 handle_prepare_lock_complete
2017-01-04 17:53:54.014168 7f28467fc700 10 librbd::ExclusiveLock: 0x7f28500187b0 handle_release_lock: r=0
```

dillaman replied:
Good catch -- I thought it looked wrong as well but adding the parens made it pass, so I just assumed it was some oddity and didn't look into it further.

dillaman commented:
@trociny update pushed

trociny commented on src/test/pybind/test_rbd.py:
```diff
@@ -1243,6 +1245,38 @@ def test_follower_write(self):
+
+                    data = rand_data(256)
+                    assert_raises(ConnectionShutdown,
+                                  blacklist_image.write, data, 0)
```
It would be nice to ensure that blacklist_on_break_lock config option is set for the image?

dillaman replied:
Sure -- we don't have any integration tests in teuthology that disable it, though, so it would be hard to get into that condition.

dillaman replied:
updated

trociny commented:
@dillaman Also observed this failure on teuthology
```
2017-01-05T09:26:28.412 INFO:tasks.workunit.client.0.smithi106.stdout:[ RUN      ] TestLibRBD.RenameViaLockOwner
2017-01-05T09:26:28.415 INFO:tasks.workunit.client.0.smithi106.stdout:using new format!
2017-01-05T09:26:28.537 INFO:tasks.workunit.client.0.smithi106.stdout:/tmp/buildd/ceph-11.1.0-6449-g1cf191e/src/test/librbd/test_librbd.cc:3505: Failure
2017-01-05T09:26:28.540 INFO:tasks.workunit.client.0.smithi106.stdout:      Expected: 0
2017-01-05T09:26:28.542 INFO:tasks.workunit.client.0.smithi106.stdout:To be equal to: rbd.rename(ioctx, name.c_str(), new_name.c_str())
2017-01-05T09:26:28.545 INFO:tasks.workunit.client.0.smithi106.stdout:      Which is: -2
2017-01-05T09:26:28.551 INFO:tasks.workunit.client.0.smithi106.stdout:[  FAILED  ] TestLibRBD.RenameViaLockOwner (145 ms)
```
http://qa-proxy.ceph.com/teuthology/trociny-2017-01-04_21:20:49-rbd-wip-mgolub-testing---basic-smithi/689831/teuthology.log

It might be not related and a known issue -- I recall we observed rename failures after socket failure injections, though I thought it was fixed.

dillaman commented:
@trociny Appended a new commit that should address the journal assertion failure due to being blacklisted

dillaman commented:
@trociny Saw that one test run [1] failed due to a race where closing the image bubbled up the ```ConnectionShutDown``` exception. I tweaked the test to ignore that exception just in case the image watcher hasn't discovered that it's been blacklisted yet.

[1] http://qa-proxy.ceph.com/teuthology/trociny-2017-01-05_20:55:31-rbd-wip-mgolub-testing---basic-smithi/692252/teuthology.log

dillaman commented:
@trociny Opened a ticket [1] against the failed ```RenameViaLockOwner``` test -- the duplicate rename requests (due to injecting socket errors) weren't being filtered out correctly.

[1] http://tracker.ceph.com/issues/18435

trociny commented on src/include/rbd/librbd.h:
```diff
@@ -337,6 +338,14 @@ CEPH_RBD_API int rbd_set_image_notification(rbd_image_t image, int fd, int type)
+                                     char **lock_owners,
+                                     size_t *max_lock_owners);
+CEPH_RBD_API int rbd_lock_get_owners_cleanup(char **lock_owners,
+                                             size_t lock_owner_count);
```
@dillaman All our other cleanup functions return void. Do you see any use in returning int instead?

trociny commented on src/test/librbd/test_librbd.cc:
```diff
@@ -4892,16 +4892,37 @@ TEST_F(TestLibRBD, ExclusiveLock)
+
+  max_lock_owners = 2;
+  ASSERT_EQ(0, rbd_lock_get_owners(image1, &lock_mode, lock_owners,
+                                   &max_lock_owners));
```
needs rbd_lock_get_owners_cleanup

dillaman commented:
@trociny The problem w/ the C version is that you would need to exclude it for unittest_librbd -- since librados_test_stub doesn't support blacklisting.

trociny commented:
@dillaman Yes, and it is excluded in my implementation.

dillaman commented:
@trociny This failure scares me that blacklisting doesn't actually work the way we expect it to (i.e. it lazily blacklists and the OSDs will eventually pick up the change but it's not guaranteed to be complete by the time the blacklist op call completes). That means when breaking the lock from an active client, we can still get stall IO written by the blacklisted client.

dillaman commented:
@trociny I addressed your comments and hopefully addressed the blacklist race condition by having the test wait for the latest OSD map and having the blacklist librados call do the same.

trociny commented on src/include/rbd/librbd.hpp:
```diff
@@ -241,6 +241,9 @@ class CEPH_RBD_API Image
   int lock_release();
+  int lock_get_owners(rbd_lock_mode_t *lock_mode,
+                      std::list<std::string> *lock_owners);
+  int lock_break(rbd_lock_mode_t lock_mode, const std::string &lock_owner);
```
@dillaman I can't think out a use case for shared lock mode, but if we plan to have this eventually, won't we want to pass several (all) owners to blacklist? Not sure how useful it is -- if the list only partially matches the current owners should we just fail?

dillaman replied:
I am keeping it generic so the end-user app can break a single shared locker if need-be.

trociny replied:
or don't blacklist (ignore lock_owner param) for shared mode?

dillaman replied:
I can build a clustering-away application with multiple instances, each using librbd shared locks. If my management layer decides to STONITH one of the shared lock owners, it should be possible.

trociny replied:
ah, never mind, I thought rados_break_lock would break the lock for all clients

trociny commented:
@ceph-jenkins retest this please
